framework/src/main/java/org/checkerframework/common/wholeprograminference/WholeProgramInferenceImplementation.java

```diff
@@ -152,11 +156,60 @@ public void updateFromMethodInvocation(
       return;
     }
 
+    // Don't infer formal parameter types from recursive calls.
+    //
+    // When performing WPI on a library, if there are no external calls (only recursive calls), then
+    // each iteration of WPI would make the formal parameter types more restrictive, leading to an
+    // infinite (or very long) loop.
+    //
+    // Consider
+    //   void myMethod(int x) { ... myMethod(x-1) ... }`
+    // On one iteration, if x has type IntRange(to=100), the recursive call's argument has type
+    // IntRange(to=99).  If that is the only call to `MyMethod`, then the formal parameter type
+    // would be updated.  On the next iteration it would be refined again to @IntRange(to=98), and
+    // so forth.  A recursive call should never restrict a formal parameter type.
+    if (isRecursiveCall(methodInvNode)) {
+      return;
+    }
+
     List<Node> arguments = methodInvNode.getArguments();
     updateInferredExecutableParameterTypes(methodElt, arguments);
     updateContracts(Analysis.BeforeOrAfter.BEFORE, methodElt, store);
   }
 
+  /**
+   * Returns true if the given call is a recursive call.
+   *
+   * @param methodInvNode a method invocation
+   * @return true if the given call is a recursive call
+   */
+  private boolean isRecursiveCall(MethodInvocationNode methodInvNode) {
+    // TODO: This tests only the name.  It should also ensure that the call isn't to an overload or
+    // a completely unrelated method of the same name.
+
+    MethodTree enclosingMethod = TreePathUtil.enclosingMethod(methodInvNode.getTreePath());
+    if (enclosingMethod == null) {
+      return false;
+    }
+
+    Tree invoked = methodInvNode.getTree().getMethodSelect();
+    Name invokedMethodName;
+    switch (invoked.getKind()) {
```

Comment:
This can be replaced with `org.checkerframework.javacutil.TreeUtils#getMethodName`.